### ISSUE TITLE
Orchestration of docker

### ISSUE THREAD

**Issue by aolney** [memphis-iis/mofacts-ies#1081]
Currently I'm not using docker compose for many of my containers. Instead I'm using `docker run` when the container does not need to interact with other containers.

Would it make sense for me to create a single docker compose with all nlp services that could then be modified for a global mofacts docker compose (with the mofacts main process, mongo, etc) OR do you want to point me to a global mofacts docker compose that I can incrementally update with said services?

**Comment by JRustyHaner** [comment]
@aolney I think we should use the second option so we can incrementally update.

**Comment by aolney** [comment]
Great, do you have such a global mofacts docker compose to start with? BTW I didn't seem to get a notification on your response, even though I subscribed to the issue; will double check my spam filter

**Comment by aolney** [comment]
I made a new repo for this: https://github.com/aolney/mofacts-automated-authoring-docker-services

There shouldn't be any problems with it, but I'm going to do some tests and report back.

Once the tests clear, I'll do a new release of `mofacts-automated-authoring`  that  has these service ports hardcoded.

**Comment by aolney** [comment]
Emailing a code bundle that uses the new docker endpoints. Suggest you replace all my previous files with these.

**Comment by aolney** [comment]
OK the docker compose repo is up to date. Some notes:

- Expect at least 40GB on disk for the images and models (haven't fully benchmarked it)
- The longformqa needs to be initialized with a data source on first deployment, after that it persists (a generic wikipedia  data source is configurable and an example for a custom data source is given; A&P text book is not distributed for copyright reasons  but can be provided via GDrive)
- The first deploy will download models from the web, after that they persist (except a few I can't get to cache)
- Expect it to take 15-30 minutes for the services to fully stand up

**Comment by JRustyHaner** [comment]
@aolney, @imrryr  I've got a compose set up for this, and have a script to build a mofacts environment in Windows that doesn't require Vagrant / Virtualbox

However, docker is not viable for local deployment on a windows system. It requires Windows Subsystem For Linux, basically Microsoft's Vagrant. We end up with the same performance problems. that Vagrant has. 

If the deploying system is inside of vagrant, we can deploy docker containers easily from there. However, Docker doesn't run effectively anymore without using Windows Subsystem for Linux.

**Comment by aolney** [comment]
I'm surprised that WSL doesn't work well since it is [one of the recommended options](https://docs.docker.com/desktop/install/windows-install/).

Have you tried the Hyper-V option? AFAIK WSL is only required for Windows Home versions - I assume most developers/researchers will have Pro/Enterprise/Education versions.

**Comment by JRustyHaner** [comment]
@aolney what's odd is that docker for windows with hyperv has been depreciated. I've looked around and most developers just say that the software isn't mature enough on windows and to stop developing in windows. (Fair enough, there's a reason we use Linux)

I'm going to try a few more things today. Usually just installing a piece of software you expect it to work, but I just found whole microsoft article on the hoops you have to jump through. I have added docker to the vagrant provisioning script, and docker works without any extra nonsense inside vagrant ( it's Linux after all ). 

A positive is that the script successfully runs meteor in windows without any virtualization, but that is only by connecting to your servers, not running the dockerized ones. This is a good step since vagrant for windows, while it works, is dreadfully slow.

The problem with docker for windows is that we have to jump through those hoops for the end user, because we don't want them to have to setup and troubleshoot docker for windows themselves. Most researchers probably would end up giving up on it. 

Tldr:

Windows Native: works and isn't slow, no docker. 

Docker for Windows: lots of hoops and setup that may not be scriptable, won't work on home editions. 

Vagrant/docker: slow on windows, but work without much fuss

Remote: vagrant/docker/meteor up: see above, but deployments work well. Meteor up can use docker repo or docker file in meteor apps directory. That's what meteor up does anyway, use a docker container.

 Meteor up also may work in native windows for deployment on a Linux server. It's just an node module after all.

**Comment by imrryr** [comment]
I don't want to add too much development time just to support Windows development when it comes right down to it. I really appreciate the way we're talking about this in detail and exploring the possibilities. Perhaps keeping it simple and in Linux is really the most efficient way to go overall especially when considering the other priorities. @aolney @JRustyHaner @MegaGeese

**Comment by JRustyHaner** [comment]
@imrryr @aolney 

I'm testing docker in Linux concurrently while I am working on windows builds on a separate workstation, so its not much of a time sink.  

A good thing actually, is that I found a workaround for windows that manually downloads docker instead of using the docker-desktop installer, and that seems to be working well.  

Since we are using docker to contain all of @aolney  's services, Meteor up actually is a good solution. It has a plugin, mup-docker-deploy, which reads the docker configuration file inside of the app's source folder (/mofactsx/) and executes it on the server. So, if the docker files that Andrew and I have been working on are valid, Meteor up would be very close to a one step deployment regardless of windows or Linux dev environments.

I've started a virtual ubuntu server on my workstation, and I'll try to deploy to it as soon as I can so I can test that method.

**Comment by JRustyHaner** [comment]
@imrryr @aolney yep, the script to install docker on windows and meteor/node dependencies works. It's just awfully slow when downloading node packages. Aparently thats a windows thing with npm. 

Thats why our windows vagrant builds are awfully slow too.
